File: python/dsa/graphs/boruvka_mst.py
# Graphs — Boruvka MST

"""Minimum spanning tree via Boruvka's algorithm.

Time: O(E log V); Space: O(V).
"""

from dsa.graphs.graph import Graph
from dsa.unionfind.union_find import UnionFind

_INF = 2**30 - 1


def total_weight(graph: Graph) -> int:
    """Returns total weight of MST.

    Time: O(E log V); Space: O(V).
    """
    n = graph.vertices()
    uf = UnionFind(n)
    total = 0
    components = n
    edges = graph.all_edges()
    while components > 1:
        best_weight = [_INF] * n
        best_to = [-1] * n
        for u, v, w in edges:
            cu = uf.find(u)
            cv = uf.find(v)
            if cu == cv:
                continue
            if w < best_weight[cu]:
                best_weight[cu] = w
                best_to[cu] = v
            if w < best_weight[cv]:
                best_weight[cv] = w
                best_to[cv] = u
        merged = False
        for c in range(n):
            if best_to[c] >= 0:
                u = c
                v = best_to[c]
                if not uf.connected(u, v):
                    uf.union(u, v)
                    total += best_weight[c]
                    components -= 1
                    merged = True
        if not merged:
            break
    return total

File: java/src/main/java/com/dsa/examples/graphs/BoruvkaMST.java
// Graphs — Boruvka MST

package com.dsa.examples.graphs;

import com.dsa.examples.unionfind.UnionFind;

import java.util.Arrays;
import java.util.List;

/**
 * Minimum spanning tree via Boruvka's algorithm.
 * Time: O(E log V); Space: O(V).
 */
public final class BoruvkaMST {

    private BoruvkaMST() {
    }

    /**
     * Returns total weight of MST.
     * Time: O(E log V); Space: O(V).
     */
    public static int totalWeight(Graph graph) {
        int n = graph.vertices();
        UnionFind uf = new UnionFind(n);
        int total = 0;
        int components = n;
        List<int[]> edges = graph.allEdges();

        while (components > 1) {
            int[] bestWeight = new int[n];
            int[] bestTo = new int[n];
            Arrays.fill(bestWeight, Integer.MAX_VALUE);
            Arrays.fill(bestTo, -1);

            for (int[] e : edges) {
                int u = e[0];
                int v = e[1];
                int w = e[2];
                int cu = uf.find(u);
                int cv = uf.find(v);
                if (cu == cv) {
                    continue;
                }
                if (w < bestWeight[cu]) {
                    bestWeight[cu] = w;
                    bestTo[cu] = v;
                }
                if (w < bestWeight[cv]) {
                    bestWeight[cv] = w;
                    bestTo[cv] = u;
                }
            }

            boolean merged = false;
            for (int c = 0; c < n; c++) {
                if (bestTo[c] >= 0) {
                    int u = c;
                    int v = bestTo[c];
                    if (!uf.connected(u, v)) {
                        uf.union(u, v);
                        total += bestWeight[c];
                        components--;
                        merged = true;
                    }
                }
            }
            if (!merged) {
                break;
            }
        }
        return total;
    }
}
